client/executor/common/src/wasmer_backend.rs

```diff
@@ -0,0 +1,433 @@
+				(Pointer::new(ptr), len)
+			};
+
+			let serialized_result_val = sandbox_context
+				.supervisor_context()
+				.read_memory(serialized_result_val_ptr, serialized_result_val_len)
+				.map_err(|_| {
+					wasmer::RuntimeError::new(
+						"Can't read the serialized result from dispatch thunk",
+					)
+				});
+
+			let deserialized_result = deallocate(
+				sandbox_context.supervisor_context(),
+				serialized_result_val_ptr,
+				"Can't deallocate memory for dispatch thunk's result",
+			)
+			.and_then(|_| serialized_result_val)
+			.and_then(|serialized_result_val| {
+				deserialize_result(&serialized_result_val)
+					.map_err(|e| wasmer::RuntimeError::new(e.to_string()))
+			})?;
+
+			if let Some(value) = deserialized_result {
+				Ok(vec![match value {
+					RuntimeValue::I32(val) => wasmer::Val::I32(val),
+					RuntimeValue::I64(val) => wasmer::Val::I64(val),
+					RuntimeValue::F32(val) => wasmer::Val::F32(val.into()),
+					RuntimeValue::F64(val) => wasmer::Val::F64(val.into()),
+				}])
+			} else {
+				Ok(vec![])
+			}
+		})
+		.expect("SandboxContextStore is set when invoking sandboxed functions; qed")
+	})
+}
+
+pub fn wasmer_new_memory(
+	context: &WasmerBackend,
+	initial: u32,
+	maximum: Option<u32>,
+) -> crate::error::Result<Memory> {
+	let ty = wasmer::MemoryType::new(initial, maximum, false);
+	let memory = Memory::Wasmer(MemoryWrapper::new(
+		wasmer::Memory::new(&context.store, ty).map_err(|_| Error::InvalidMemoryReference)?,
+	));
+
+	Ok(memory)
+}
+
+/// In order to enforce memory access protocol to the backend memory
+/// we wrap it with `RefCell` and encapsulate all memory operations.
+#[derive(Debug, Clone)]
+pub struct MemoryWrapper {
+	buffer: Rc<RefCell<wasmer::Memory>>,
+}
+
+impl MemoryWrapper {
+	/// Take ownership of the memory region and return a wrapper object
+	pub fn new(memory: wasmer::Memory) -> Self {
+		Self { buffer: Rc::new(RefCell::new(memory)) }
+	}
+
+	/// Returns linear memory of the wasm instance as a slice.
+	///
+	/// # Safety
+	///
+	/// Wasmer doesn't provide comprehensive documentation about the exact behavior of the data
+	/// pointer. If a dynamic style heap is used the base pointer of the heap can change. Since
+	/// growing, we cannot guarantee the lifetime of the returned slice reference.
+	unsafe fn memory_as_slice(memory: &wasmer::Memory) -> &[u8] {
+		let ptr = memory.data_ptr() as *const _;
+		let len: usize = memory.data_size().try_into().expect("data size should fit into usize");
+
+		if len == 0 {
+			&[]
+		} else {
+			core::slice::from_raw_parts(ptr, len)
+		}
+	}
+
+	/// Returns linear memory of the wasm instance as a slice.
+	///
+	/// # Safety
+	///
+	/// See `[memory_as_slice]`. In addition to those requirements, since a mutable reference is
+	/// returned it must be ensured that only one mutable and no shared references to memory
+	/// exists at the same time.
+	unsafe fn memory_as_slice_mut(memory: &wasmer::Memory) -> &mut [u8] {
```

Comment:
This converts a immutable reference to a mutable slice as pointed out by clippy. The memory argument should be mutable.